core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java

```diff
@@ -47,12 +47,15 @@ public Object getPropertyValue(@Nonnull Property property) {
   }
 
   @Override
-  public void setPropertyValue(@Nonnull Property property, Object value) {
+  public void setPropertyValue(@Nonnull Property property, @Nullable Object value) {
     Objects.requireNonNull(property, "Property should not be null");
     if (!getClassifier().allProperties().contains(property)) {
       throw new IllegalArgumentException(
           "Property " + property + " is not belonging to classifier " + getClassifier());
     }
+    if (property.getID() == null) {
```

Comment:
Keys have the particularity of being stable across versions, but besides that my understanding is that they have no other special role. In general, both the key and the id should be defined and be unique within the features of a node, so both could be used as keys in a map of features. I tend to use IDs and not keys precisely because IDs are unique _also across versions_, so they permit the identification of features more uniquely. 

In this particular case, either the id or the key would work. I think there is a slight advantage in using the id because I could represent a state in which the node is being migrated, and for an instant, it has values for the same property in two different versions of the language. In a similar situation I could store both the value for the "old" property and the "new" property, while if I were using the key I would not be able to recognize the case. This seems admittedly a corner case.

In other words, I do not think the specs indicate that internally we should use organize features by ids or keys, as both should be unique. I would tend to use ids and not keys because I think they can be (very marginally) better because they are more unique, and because when in doubt I would let the status quo prevail. Does it make sense?

I can definitely move the check up.

Reply:
Disclaimer: I think the key vs. id debate has a large "philosophical" part, so take my arguments in that view.

In another (very corner) case one could imagine a hard-coded language that does not have ids for its elements -- we don't require language elements to have ids (kind-of, I'm pretty sure all implementations still require it to keep the "everything is a node" mantra).

I think it gets a lot more interesting when we start actually using the _version_ part of a meta-pointer: `language` (key) + `version` (of the language) uniquely identify the language; within that, the `key` uniquely identifies the element within the language. 
Element key and language key have to match perfectly, otherwise they don't make sense.

Version is a different matter. We explicitly separated it from language key to allow some "softness" in matching. Example: In my system I only know language `A` in version `x` (written as `A@x`), but try to deserialize an instance of `A` in version `y`. Then I could still use the language and elements of `A@x` and either be lucky that they are compatible, or at least deserialize as much as possible.

I imagine then it might get important to use keys instead of ids, because they might be applicable to the same language in a different version -- which is the desired effect. It's kind of the same argument as you have, but with a very different interpretation (-:

Reply:
On the key vs. id debate I agree it can get "philosophical" and I think it has impact in at least two points:
- In this class
- In PrimitiveValuesSerialization

In this class, I do not feel strongly, so if you prefer we use keys instead of IDs, that works for me. I have implemented this change in the last commit.

In PrimitiveValuesSerialization instead I feel strongly, as I think that we may want to specify different ways to serialize different versions of a certain primitive type. So there I use ids and not keys.